src/stats/conversions.ts

```diff
@@ -1,221 +1,216 @@
-    const actionCounter = playerFrame.actionStateCounter;
-    const prevActionCounter = prevPlayerFrame.actionStateCounter;
-    const actionFrameCounterReset = actionCounter < prevActionCounter;
-    if (actionChangedSinceHit || actionFrameCounterReset) {
-      state.lastHitAnimation = null;
-    }
+import { FrameEntryType, FramesType, MoveLandedType, ConversionType, PlayerIndexedType, Frames } from "./common";
+import { isDamaged, isGrabbed, calcDamageTaken, isInControl, didLoseStock, Timers } from "./common";
+import { StatComputer } from './stats';
+
+interface PlayerConversionState {
+  conversion: ConversionType | null;
+  move: MoveLandedType | null;
+  resetCounter: number;
+  lastHitAnimation: number | null;
+}
 
-    // If opponent took damage and was put in some kind of stun this frame, either
-    // start a conversion or
-    if (opntIsDamaged || opntIsGrabbed) {
-      if (!state.conversion) {
-        state.conversion = {
-          playerIndex: indices.playerIndex,
-          opponentIndex: indices.opponentIndex,
-          startFrame: playerFrame.frame,
-          endFrame: null,
-          startPercent: prevOpponentFrame.percent || 0,
-          currentPercent: opponentFrame.percent || 0,
-          endPercent: null,
-          moves: [],
-          didKill: false,
-          openingType: "unknown", // Will be updated later
-        };
+interface MetadataType {
+  lastEndFrameByOppIdx: {
+    [oppIdx: number]: number;
+  };
+}
 
-        conversions.push(state.conversion);
+export class ConversionComputer implements StatComputer<ConversionType[]> {
+  private opponentIndices: PlayerIndexedType[];
+  private conversions: ConversionType[] = [];
+  private state: Map<PlayerIndexedType, PlayerConversionState>;
+  private metadata: MetadataType;
+
+  public constructor(opponentIndices: PlayerIndexedType[]) {
+    this.opponentIndices = opponentIndices;
+    this.state = new Map<PlayerIndexedType, PlayerConversionState>();
+    this.metadata = {
+      lastEndFrameByOppIdx: {},
+    };
+
+    this.opponentIndices.forEach((indices) => {
+      const playerState: PlayerConversionState = {
+        conversion: null,
+        move: null,
+        resetCounter: 0,
+        lastHitAnimation: null,
       }
-
-      if (opntDamageTaken) {
-        // If animation of last hit has been cleared that means this is a new move. This
-        // prevents counting multiple hits from the same move such as fox's drill
-        if (!state.lastHitAnimation) {
-          state.move = {
-            frame: playerFrame.frame,
-            moveId: playerFrame.lastAttackLanded,
-            hitCount: 0,
-            damage: 0,
-          };
-
-          state.conversion.moves.push(state.move);
-        }
-
-        if (state.move) {
-          state.move.hitCount += 1;
-          state.move.damage += opntDamageTaken;
+      this.state.set(indices, playerState);
+    })
+  }
+
+  public processFrame(frame: FrameEntryType, allFrames: FramesType): void {
```

Comment:
Something I liked about my method that I had been working on is that in my equivalent to the "calculator" class, I had abstracted out the concept of the alternating player indices completely. Think it made for simpler-to-implement interface for the type that is most extensible in the system.

Reply:
Perhaps this could be abstracted out via class inheritance somehow?